src/mlpack/core/optimizers/cmaes/cmaes_impl.hpp

```diff
@@ -0,0 +1,942 @@
+    xmean[0] =N;
+    ++xmean;
+    xold = new T[N+2];
+    xold[0] = N;
+    ++xold;
+    xBestEver = new T[N+3];
+    xBestEver[0] = N;
+    ++xBestEver;
+    xBestEver[N] = std::numeric_limits<T>::max();
+    output = new T[N+2];
+    output[0] = N;
+    ++output;
+    rgD = new T[N];
+    C = new T*[N];
+    B = new T*[N];
+    publicFitness = new T[lambda];
+    functionValues = new T[lambda+1];
+    functionValues[0] = lambda;
+    ++functionValues;
+    const int historySize = 10 + (int) ceil(3.*10.*N/lambda);
+    funcValueHistory = new T[historySize + 1];
+    funcValueHistory[0] = (T) historySize;
+    funcValueHistory++;
+
+    for (int i = 0; i <N; ++i)
+    {
+      C[i] = new T[i+1];
+      B[i] = new T[N];
+    }
+    index = new int[lambda];
+    for (int i = 0; i <lambda; ++i)
+        index[i] = i;
+    population = new T*[lambda];
+    for (int i = 0; i <lambda; ++i)
+    {
+      population[i] = new T[N+2];
+      population[i][0] = N;
+      population[i]++;
+      for (int j = 0; j <N; j++)
+        population[i][j] = 0.0;
+    }
+
+    for (int i = 0; i <lambda; i++)
+    {
+      functionValues[i] = std::numeric_limits<T>::max();
+    }
+    for (int i = 0; i < historySize; i++)
+    {
+      funcValueHistory[i] = std::numeric_limits<T>::max();
+    }
+    for (int i = 0; i <N; ++i)
+      for (int j = 0; j < i; ++j)
+        C[i][j] = B[i][j] = B[j][i] = 0.;
+
+    for (int i = 0; i <N; ++i)
+    {
+      B[i][i] = T(1);
+      C[i][i] = rgD[i] = rgInitialStds[i]*std::sqrt(N/trace);
+      C[i][i] *= C[i][i];
+      pc[i] = ps[i] = T(0);
+    }
+    minEW = minElement(rgD,N);
+    minEW = minEW*minEW;
+    maxEW = maxElement(rgD,N);
+    maxEW = maxEW*maxEW;
+
+    maxdiagC = C[0][0];
+    for (int i = 1; i < N; ++i) if (maxdiagC < C[i][i]) maxdiagC = C[i][i];
+    mindiagC = C[0][0];
+    for (int i = 1; i < N; ++i) if (mindiagC > C[i][i]) mindiagC = C[i][i];
+
+    for (int i = 0; i <N; ++i)
+      xmean[i] = xold[i] = xstart[i];
+    
+    if (typicalXcase)
+      for (int i = 0; i < N; ++i)
+        xmean[i] += sigma*rgD[i]*rand.gauss();
+
+    stStopFitness.flg = false;
+    stStopFitness.val = -std::numeric_limits<T>::max();
+
+    return publicFitness;
+  
+  }
+
+//! Optimize the function (minimize).
+template<typename funcType, typename T>
+T CMAES<funcType, T>::Optimize(funcType& func, T *iterate)
+{
+  // numof functions for the N and the evaluate for the function
+  // return the final answer  and the xfinal in the iterate itself
+  //TEST FOR TERMINATE SAMPLE AND UPDATE
+
+// Iterate until stop criterion holds
+  while(!testForTermination())
+  { 
+    double *const*pop;
+    // Generate lambda new search points, sample population
+    pop = samplePopulation();
+
+    //const size_t numFunctions = func.NumFunctions();
+
+    // evaluate the new search points using evaluate function given by the user
+    for (int i = 0; i < lambda; ++i)
+     functionFitness[i] = func.Evaluate(pop[i]);
```

Comment:
so for a user in other algorithms he has to make all the test cases (where each case is a dimension) and then one by one it is taken using index value.. this can be tedius for user with large equations. cmaes just need the total equation. but when this is implemented 
the standard way of evaluate changes leading to unmatched parameter for other algos lets take sgd ..
which one should i go for ?